pallets/validators-manager/src/lib.rs

```diff
@@ -629,37 +567,344 @@ impl<T: Config> Pallet<T> {
+    }
+
+    fn handle_registration_failure(
+        account_id: &T::AccountId,
+        ingress_counter: IngressCounter,
+        reason: &str,
+        cleanup_deposit: bool,
+    ) {
+        log::error!("Validator registration failed for {:?}: {}", account_id, reason);
+
+        Self::cleanup_registration_storage(&account_id, ingress_counter, cleanup_deposit);
+
+        Self::deposit_event(Event::<T>::ValidatorRegistrationFailed {
+            validator_id: account_id.clone(),
+            reason: reason.as_bytes().to_vec(),
+        });
+    }
+
+    fn handle_deregistration_failure(
+        account_id: &T::AccountId,
+        ingress_counter: IngressCounter,
+        reason: &str,
+    ) {
+        log::error!("Validator deregistration failed for {:?}: {}", account_id, reason);
+
+        <ValidatorActions<T>>::remove(&account_id, ingress_counter);
+
+        Self::deposit_event(Event::<T>::ValidatorDeregistrationFailed {
+            validator_id: account_id.clone(),
+            reason: reason.as_bytes().to_vec(),
+        });
+    }
+
+    /// Rollback and cleanup state when T1 operation fails
+    fn rollback_failed_validator_action(
+        account_id: &T::AccountId,
+        ingress_counter: IngressCounter,
+        action_type: ValidatorsActionType,
+        tx_id: EthereumId,
+    ) {
+        // Type-specific cleanup
+        if action_type.is_registration() {
+            Self::cleanup_registration_storage(&account_id, ingress_counter, true);
+        } else {
+            // For non-registration actions, just remove the validator action entry
+            <ValidatorActions<T>>::remove(&account_id, ingress_counter);
+        }
+
+        Self::deposit_event(Event::<T>::ValidatorActionFailedOnEthereum { tx_id });
+    }
 }
 
 impl<T: Config> BridgeInterfaceNotification for Pallet<T> {
     fn process_result(tx_id: u32, caller_id: Vec<u8>, succeeded: bool) -> DispatchResult {
-        // TODO: Update data structure to use tx_id as key
-        if caller_id == PALLET_ID.to_vec() {
-            if succeeded {
-                log::info!(
-                    "✅  Transaction with ID {} was successfully published to Ethereum.",
-                    tx_id
-                );
-                Self::deposit_event(Event::<T>::PublishingValidatorActionOnEthereumSucceeded {
-                    tx_id,
-                });
-            } else {
-                log::error!("❌ Transaction with ID {} failed to publish to Ethereum.", tx_id);
-                Self::deposit_event(Event::<T>::PublishingValidatorActionOnEthereumFailed {
-                    tx_id,
-                });
+        if caller_id != PALLET_ID.to_vec() {
+            return Ok(())
+        }
+
+        // Find the ValidatorActions entry with matching tx_id
+        let mut found_entry: Option<(T::AccountId, IngressCounter, ValidatorsActionType)> = None;
+
+        for (account_id, ingress_counter, validators_action_data) in <ValidatorActions<T>>::iter() {
```

Comment:
I would prefer to bring back the separate storage item to keep track of `tx_id` -> `Validator action`. It saves iterating and also allows mapping the tx_id we get on the new events to an action.